### PR TITLE
Refactor making uploading from forms more dynamic

### PR DESCRIPTION
**Pull request by RemcoSimonides**
To do:
- [x] Upload images to Firestorage when ProfileForm, OrganizationForm or MovieForm is submitted

### CODE REVIEW

File: apps/festival/festival/src/app/dashboard/tunnel/movie-tunnel.component.ts [hunk 1]
```diff
@@ -103,53 +109,30 @@ export class MovieTunnelComponent implements TunnelRoot, OnInit {
...
-      }
-
-      this.storage.ref(newRef).put(blob);
-      imageForm.ref.setValue(newRef);
-    }
-
-    imageForm.blob.setValue('');
-    imageForm.newRef.setValue('');
-    imageForm.delete.setValue(false);
+  private removePaths(movieData: Movie) {
```
Comment (GrandSchtroumpf) [inline apps/festival/festival/src/app/dashboard/tunnel/movie-tunnel.component.ts:125]: You need to remove this.

File: apps/festival/festival/src/app/dashboard/tunnel/movie-tunnel.component.ts [hunk 2]
```diff
@@ -103,53 +109,30 @@ export class MovieTunnelComponent implements TunnelRoot, OnInit {
...
+    const promotionalElements = movieData.promotionalElements;
+    this.remove(promotionalElements.banner.media)
+    promotionalElements.poster.forEach(poster => this.remove(poster.media))
+    promotionalElements.still_photo.forEach(still_photo => this.remove(still_photo.media))
+  }
 
+  private remove(mediaData) {
```
Comment (GrandSchtroumpf) [inline apps/festival/festival/src/app/dashboard/tunnel/movie-tunnel.component.ts:132]: Also remove please

File: libs/movie/src/lib/movie/form/promotional-elements/promotional-elements.form.ts [hunk 3]
```diff
@@ -74,4 +74,20 @@ export class MoviePromotionalElementsForm extends FormEntity<MoviePromotionalEle
   constructor(promotionalElements?: MoviePromotionalElements) {
     super(createMoviePromotionalElementsControls(promotionalElements));
   }
+
+  get mediaPaths(): string[] {
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/form/promotional-elements/promotional-elements.form.ts:78]: Put that either in : 
- media.model.ts
- movie.model.ts

```typescript
export const movieMediaPaths = [
  'promotionalElements',
];
```

Comment (GrandSchtroumpf) [reply libs/movie/src/lib/movie/form/promotional-elements/promotional-elements.form.ts:78]: Or you could do : 
```typescript
function searchForMediaRef(object: any) {
  const refs = [];
  for (const key in object) {
    if (isImgRef(object[key])) {
      refs.push(object[key]); 
      delete object[key];
    } 
    if (Array.isArray(object[key]) {
      // handle array
    }
    if (typeof object[key] === object && !!object[key]) { // null is an object
       const childRefs = searchForMediaRef(object[key]);
       refs.concat(childRefs);
    }
  }
  return refs;
}
```

```typescript
const movie = movieForm.value;
const refs = searchForMediaRef(movie);
// movie keys which have references have been removed
```

File: libs/utils/src/lib/media/media.service.ts [hunk 4]
```diff
@@ -29,4 +33,48 @@ export class ImageUploader {
     }
   }
 
+  public extractMediaForm(form, extractPaths: string[]) {
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/media/media.service.ts:36]: When you don't use `this` in a method, create a pure function outside of the class instead.

File: libs/utils/src/lib/media/media.service.ts [hunk 5]
```diff
@@ -33,48 +33,70 @@ export class ImageUploader {
     }
   }
 
-  public extractMediaForm(form, extractPaths: string[]) {
-    
-    const value = form.value;
-    const extracted = {};
 
-    extractPaths.forEach(path => {
+  public handleMediaForms(form) {
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/media/media.service.ts:37]: Find a better name that express what this method is doing.
And type the parameter.

File: apps/festival/festival/src/app/dashboard/tunnel/movie-tunnel.component.ts [hunk 6]
```diff
@@ -81,11 +81,10 @@ export class MovieTunnelComponent implements TunnelRoot, OnInit {
   // Should save movie
   public async save() {
```
Comment (GrandSchtroumpf) [inline apps/festival/festival/src/app/dashboard/tunnel/movie-tunnel.component.ts:83]:
```typescript
// Remove the medias fields from the movie
const [ value, media ] = extractMedia(this.form.value);

// Upload the media
this.mediaService.upload(media);

// Save the movie without media
const movie: Movie = mergeDeep(this.query.getActive(), value);
this.movieService.upload(movie);
```

File: libs/utils/src/lib/media/media.service.ts [hunk 7]
```diff
@@ -33,48 +33,70 @@ export class ImageUploader {
...
+    return formValue;
+  
+  }
+
+}
 
+function searchForMediaRef(object: any): ImgRefForm[] {
+  let imgRefs = [];
+
+  if ("controls" in object) {
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/media/media.service.ts:85]: Object is a value, not a form.

File: apps/festival/festival/src/app/dashboard/tunnel/movie-tunnel.component.ts (hunk 6 above)
Comment (GrandSchtroumpf) [reply apps/festival/festival/src/app/dashboard/tunnel/movie-tunnel.component.ts:83]:
```typescript
function extractMedia(origin: object) {
  const media = [];
  const value = Object.assign({}, origin); // copy the origin
  // look for each media field
  // Add to media
  // Remove field from value
  return [ value, media ];
}
```

File: apps/festival/festival/src/app/dashboard/tunnel/movie-tunnel.component.ts [hunk 8]
```diff
@@ -76,10 +80,14 @@ export class MovieTunnelComponent implements TunnelRoot, OnInit {
 
   // Should save movie
   public async save() {
-    const movie: Movie = mergeDeep(this.query.getActive(), this.form.value);
-    // This is handled by a backend function
-    delete movie.promotionalElements; // @TODO (##2987)
-    await this.service.update(movie);
+
+    const value = Object.assign({}, this.form.value);
+    const media = extractMedia(value);
```
Comment (GrandSchtroumpf) [inline apps/festival/festival/src/app/dashboard/tunnel/movie-tunnel.component.ts:85]: As describe in the previous comment I think it's better to do the `Object.assign({}, origin)` in the `extractMedia` :
```typescript
const [ movie, media ] = extractMedia(this.form.value);
```

File: libs/ui/src/lib/media/cropper/cropper/cropper.component.ts [hunk 9]
```diff
@@ -124,24 +120,20 @@ export class CropperComponent implements OnDestroy, OnInit {
...
       const blob = b64toBlob(this.croppedImage);
 
-      this.percentage$ = this.ref.put(blob).percentageChanges().pipe(
-        finalize(async () => {
-          this.nextStep('upload_complete')
-        })
-      );
+      this.nextStep('show');
+
+      this.form.newRef.setValue(`${this.storagePath}/original/${this.fileName}`);
```
Comment (GrandSchtroumpf) [inline libs/ui/src/lib/media/cropper/cropper/cropper.component.ts:134]: You should be able to do : 
```typescript
this.form.patchValue({
  newRef: `${this.storagePath}/original/${this.fileName}`,
  delete: false,
  blob,
})
```

File: libs/ui/src/lib/media/cropper/cropper/cropper.component.ts [hunk 10]
```diff
@@ -153,12 +145,14 @@ export class CropperComponent implements OnDestroy, OnInit {
...
-      this.storage.storage.refFromURL(path).delete();
-      this.form.reset();
-      this.nextStep('drop');
-    });
-    this.sub.add(deleteSubscription$);
+    if (this.croppedImage) {
+      this.croppedImage = '';
+    }
+
+    this.form.delete.setValue(true);
```
Comment (GrandSchtroumpf) [inline libs/ui/src/lib/media/cropper/cropper/cropper.component.ts:152]: Same as above;

File: apps/festival/festival/src/app/dashboard/tunnel/summary/summary.component.ts [hunk 11]
```diff
@@ -38,7 +41,9 @@ export class TunnelSummaryComponent {
 
   public async submit() {
     if (this.form.valid) {
-      const movie: Movie = mergeDeep(this.query.getActive(), this.form.value);
+      const [ value, media ] = extractMedia(this.form.value);
+      this.media.uploadOrDeleteMedia(media);
```
Comment (fritzschoff) [inline apps/festival/festival/src/app/dashboard/tunnel/summary/summary.component.ts:45]: you want to await it? not sure if this func is async

File: libs/movie/src/lib/movie/form/promotional-elements/promotional-elements.form.ts [hunk 12]
```diff
@@ -74,4 +74,20 @@ export class MoviePromotionalElementsForm extends FormEntity<MoviePromotionalEle
...
+    const paths = [];
```
Comment (fritzschoff) [inline libs/movie/src/lib/movie/form/promotional-elements/promotional-elements.form.ts:79]: maybe type it with `ImgRef`  if it is this type..

File: libs/ui/src/lib/media/cropper/cropper/cropper.component.html [hunk 13]
```diff
@@ -41,7 +41,7 @@
   <article class="show" *ngSwitchCase="'show'" @finalZoom>
     <ng-container>
       <div class="show">
-        <img class="croppedImage" [src]="previewUrl | async" />
+        <img class="croppedImage" [src]="croppedImage || (previewUrl | async)" />
```
Comment (fritzschoff) [inline libs/ui/src/lib/media/cropper/cropper/cropper.component.html:44]: previewUrl is Observable so add `$` for extra money ;)

File: libs/ui/src/lib/media/cropper/cropper/cropper.component.ts [hunk 14]
```diff
@@ -68,7 +70,7 @@ export class CropperComponent implements OnDestroy, OnInit {
   step$ = this.step.asObservable();
   file: File;
   croppedImage: string;
-  cropRatio: string;
+  cropRatio: number;
   parentWidth: number;
   prev: CropStep;
   previewUrl: Observable<string | null>;
```
Comment (fritzschoff) [inline libs/ui/src/lib/media/cropper/cropper/cropper.component.ts:76]: as well here to fit the naming convention

File: apps/festival/festival/src/app/dashboard/tunnel/summary/summary.component.ts (hunk 11 above)
Comment (RemcoSimonides) [reply apps/festival/festival/src/app/dashboard/tunnel/summary/summary.component.ts:45]: No, there is no need to wait for the images to be uploaded or deleted here. We have your upload component for this :)

File: libs/movie/src/lib/movie/form/promotional-elements/promotional-elements.form.ts (hunk 12 above)
Comment (RemcoSimonides) [reply libs/movie/src/lib/movie/form/promotional-elements/promotional-elements.form.ts:79]: Actually this whole getter isn't used anymore. I am removing it

File: libs/utils/src/lib/pipes/reference-path.pipe.ts [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+        poster: (id: string, i: number) => `movies/${id}/promotionalElements.poster.${i}.media`, 
+        still: (id: string, i: number) => `movies/${id}/promotionalElements.still_photo.${i}.media`,
+    },
+    profile: {
+        avatar: (uid: string) => `users/${uid}/avatar`
+    },
+    organization: {
+      logo: (id: string) => `orgs/${id}/logo`
+    }
+}
```
Comment (pldespaigne) [inline libs/utils/src/lib/pipes/reference-path.pipe.ts:15]: indentation is not coherent 😝 really cool pipe other wise 👍

File: libs/utils/src/lib/media/media.model.ts [hunk 16]
```diff
@@ -1 +1,33 @@
...
+export * from './media.firestore';
+
+export function extractMedia(origin: any) {
+  const value = Object.assign({}, origin);
+  const media = extractMediaValue(value);
+  return [value, media];
+}
+
+function extractMediaValue(value) {
+  let media = [];
```
Comment (pldespaigne) [inline libs/utils/src/lib/media/media.model.ts:10]: I think this was causing the type error in the CI

File: apps/catalog/catalog/src/app/dashboard/movie-tunnel/media-image/media-image.component.html [hunk 17]
```diff
@@ -48,7 +48,7 @@ <h3 type>Poster</h3>
         <ng-container *ngFor="let control of stillPhoto.controls; let i = index; let last = last">
           <span class="still mat-body-2">Still # {{i + 1}}</span>
           <drop-cropper
-            [storagePath]="stillPath"
+            [storagePath]="'movies/' + movie.id + '/promotionalElements.still_photo.' + i + '.media'"
```
Comment (GrandSchtroumpf) [inline apps/catalog/catalog/src/app/dashboard/movie-tunnel/media-image/media-image.component.html:51]: Why not using the pipe you created here ?

File: apps/catalog/catalog/src/app/dashboard/movie-tunnel/movie-tunnel.component.ts [hunk 18]
```diff
@@ -87,7 +91,9 @@ export class MovieTunnelComponent implements TunnelRoot, OnInit {
 
   // Should save movie
   public async save() {
-    const movie: Movie = mergeDeep(this.query.getActive(), this.form.value);
+    const [ value, media ] = extractMedia(this.form.value);
+    this.media.uploadOrDeleteMedia(media);
+    const movie: Movie = mergeDeep(this.query.getActive(), value);
```
Comment (GrandSchtroumpf) [inline apps/catalog/catalog/src/app/dashboard/movie-tunnel/movie-tunnel.component.ts:96]: I think this is dangerous to do that after extracting media as you're actually adding back the current media...
I think you ned to put it before `extractMedia`

File: apps/catalog/catalog/src/app/dashboard/movie-tunnel/summary/summary.component.ts [hunk 19]
```diff
@@ -43,7 +46,9 @@ export class TunnelSummaryComponent {
 
   public async submit() {
     if (this.form.valid) {
-      const movie: Movie = mergeDeep(this.query.getActive(), this.form.value);
+      const [ value, media ] = extractMedia(this.form.value);
+      this.media.uploadOrDeleteMedia(media);
+      const movie: Movie = mergeDeep(this.query.getActive(), value);
```
Comment (GrandSchtroumpf) [inline apps/catalog/catalog/src/app/dashboard/movie-tunnel/summary/summary.component.ts:51]: Same thing here.

Comment (GrandSchtroumpf) [reply apps/catalog/catalog/src/app/dashboard/movie-tunnel/summary/summary.component.ts:51]: **Todo in another PR**: 
maybe you should do all this logic in the `sav` method from MovieService so you don't repeat yourself.

File: apps/catalog/catalog/src/app/dashboard/movie-tunnel/media-image/media-image.component.html (hunk 17 above)
Comment (RemcoSimonides) [reply apps/catalog/catalog/src/app/dashboard/movie-tunnel/media-image/media-image.component.html:51]: Good catch! It probably changed during a merge

File: libs/utils/src/lib/media/media.model.ts [hunk 20]
```diff
@@ -1 +1,29 @@
...
+  const value = Object.assign({}, origin);
+  const media = extractMediaValue(value);
+  return [value, media];
+}
+
+function extractMediaValue(value: any) {
+  let media: ImgRef[] = [];
+  for (const key in value) {
+    if (isMedia(value[key])) {
+      if (!!value[key]) {
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/media/media.model.ts:15]: I'm not sure to understand how `value[key]` can be a media & not been truthy. For me if looks like you don't need this.
```typescript
for (const key in value) {
  const field = value[key];
  if (field) {
    if (isMedia(field)) {
      media.push(field);
      delete value[key]:
    } else if (typeof value[key] === 'object') {
      const childMedia = extractMediaValue(value[key]);
      media = media.concat(childMedia);
    }
  }
}
```

File: libs/utils/src/lib/media/media.service.ts [hunk 21]
```diff
@@ -152,4 +156,23 @@ export class MediaService {
       this.overlayRef.attach(new ComponentPortal(UploadWidgetComponent));
     }
   }
+
+  uploadOrDeleteMedia(medias: ImgRef[]) {
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/media/media.service.ts:160]: In another PR we should improve this method to include some try / catch. I lt you open an issue for that.

File: libs/utils/src/lib/media/media.model.ts [hunk 22]
```diff
@@ -1 +1,29 @@
...
+export * from './media.firestore';
+
+export function extractMedia(origin: any) {
+  const value = Object.assign({}, origin);
+  const media = extractMediaValue(value);
+  return [value, media];
+}
+
+function extractMediaValue(value: any) {
+  let media: ImgRef[] = [];
```
[the remark this answers is not in the document]
Comment (RemcoSimonides) [reply libs/utils/src/lib/media/media.model.ts:12]:
```typescript
media = media.concat(childMedia);
```
concat functions returns a new array and doesn't mutate the passed in one, so I need to reassign media

[the remark this answers is not in the document]
Comment (GrandSchtroumpf) [reply libs/utils/src/lib/media/media.model.ts:12]: ok my bad.

File: libs/utils/src/lib/media/media.service.ts (hunk 21 above)
Comment (RemcoSimonides) [reply libs/utils/src/lib/media/media.service.ts:160]: Could you elaborate? What should happen when the upload or delete fails? The upload component which shows the upload status will show an error I expect. Or should I just add a simple try catch block for error logging?

**Pull request merged**